src/main/cpp/RLBotInterface/BallPrediction/inc/mat.h

```diff
@@ -0,0 +1,350 @@
+    }
+
+    float operator()(const int i0, const int i1) const{
+      return data[i0 + N0 * i1];
+    }
+
+};
+
+template < int m, int n >
+inline mat < n, m > transpose(const mat < m, n > & A) {
+  mat < n, m > AT;
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      AT(j,i) = A(i,j);
+    }
+  }
+  return AT;
+}
+
+template < int m, int n >
+inline mat < m, n > operator+(const mat < m, n > & A,
+                                   const mat < m, n > & B) {
+  mat < m, n > C;
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      C(i,j) = A(i,j) + B(i,j);
+    }
+  }
+  return C;
+}
+
+template < int m, int n >
+inline mat < m, n > operator*(const mat < m, n > & A,
+                                   const float other) {
+  mat < m, n > B;
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      B(i,j) = other * A(i,j);
+    }
+  }
+  return B;
+}
+
+template < int m, int n >
+inline mat < m, n > operator*(const float other,
+                                   const mat < m, n > & A) {
+  mat < m, n > B;
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      B(i,j) = other * A(i,j);
+    }
+  }
+  return B;
+}
+
+template < int n >
+inline mat < n, n > eye(){
+  mat < n, n > I;
+  for (int i = 0; i < n; i++) {
+    for (int j = 0; j < n; j++) {
+      I(i,j) = (i == j);
+    }
+  }
+  return I;
+}
+
+template < int m, int n >
+inline float fnorm(const mat < m, n > & A) {
+  float sum = 0.0;
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      sum += A(i,j);
+    }
+  }
+  return sqrt(sum);
+}
+
+template < int n >
+inline float tr(const mat < n, n > & A) {
+  float sum = 0.0;
+  for (int i = 0; i < n; i++) {
+    sum += A(i,i);
+  }
+  return sum;
+}
+
+inline float det(const mat < 2, 2 > & A){
+  return A(0,0)*A(1,1)-A(0,1)*A(1,0);
+}
+
+inline float det(const mat < 3, 3 > & A){
+  return A(0,0)*A(1,1)*A(2,2)+A(0,1)*A(1,2)*A(2,0)+
+         A(0,2)*A(1,0)*A(2,1)-A(0,0)*A(1,2)*A(2,1)-
+         A(0,1)*A(1,0)*A(2,2)-A(0,2)*A(1,1)*A(2,0);
+}
+
+inline float det(const mat < 4, 4 > & A){
+  return A(0,3)*A(1,2)*A(2,1)*A(3,0)-A(0,2)*A(1,3)*A(2,1)*A(3,0)-
+         A(0,3)*A(1,1)*A(2,2)*A(3,0)+A(0,1)*A(1,3)*A(2,2)*A(3,0)+
+         A(0,2)*A(1,1)*A(2,3)*A(3,0)-A(0,1)*A(1,2)*A(2,3)*A(3,0)-
+         A(0,3)*A(1,2)*A(2,0)*A(3,1)+A(0,2)*A(1,3)*A(2,0)*A(3,1)+
+         A(0,3)*A(1,0)*A(2,2)*A(3,1)-A(0,0)*A(1,3)*A(2,2)*A(3,1)-
+         A(0,2)*A(1,0)*A(2,3)*A(3,1)+A(0,0)*A(1,2)*A(2,3)*A(3,1)+
+         A(0,3)*A(1,1)*A(2,0)*A(3,2)-A(0,1)*A(1,3)*A(2,0)*A(3,2)-
+         A(0,3)*A(1,0)*A(2,1)*A(3,2)+A(0,0)*A(1,3)*A(2,1)*A(3,2)+
+         A(0,1)*A(1,0)*A(2,3)*A(3,2)-A(0,0)*A(1,1)*A(2,3)*A(3,2)-
+         A(0,2)*A(1,1)*A(2,0)*A(3,3)+A(0,1)*A(1,2)*A(2,0)*A(3,3)+
+         A(0,2)*A(1,0)*A(2,1)*A(3,3)-A(0,0)*A(1,2)*A(2,1)*A(3,3)-
+         A(0,1)*A(1,0)*A(2,2)*A(3,3)+A(0,0)*A(1,1)*A(2,2)*A(3,3);
+}
+
+template < int n >
+inline float I1(const mat < n, n > & A) {
```

Comment:
same here with more context to name

Reply:
Sorry, these aren't used in RL stuff, really. I1, I2, and I3 are another commonly used convention for the invariants of a matrix.